### PR TITLE
Add the content of third_party.cmake, test=develop

### PR DESCRIPTION
**Pull request by zhwesky2010**
Add the content of file third_party.cmake, these file (cares.cmake/cub.cmake/dlpack.cmake/eigen.cmake/grpc.cmake/libmct.cmake/pslib.cmake/pslib_brpc.cmake/rocprim.cmake/threadpool.cmake/xbyak.cmake/xxhash.cmake) have some code to moved to third_party.cmake.

These CODE has been moved to third_party.cmake:

- 1.The flag that control whether this submodule should be compiled;
- 2.Error message of submodule when the flag is wrong;

In addition, lisense information has been added.

This PR follow up #20887 .

### CODE REVIEW

File: cmake/external/pslib_brpc.cmake [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2017 PaddlePaddle Authors. All Rights Reserved.
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (liupluswei) [inline cmake/external/pslib_brpc.cmake:1]: we have different versions of copyright, can we unify this?

File: cmake/third_party.cmake [hunk 2]
```diff
@@ -125,9 +148,16 @@ if(WITH_GPU)
...
 
 if(WITH_PSLIB)
-    include(external/libmct)
-    include(external/pslib)
-    include(external/pslib_brpc)
+    include(external/pslib)          # download, build, install pslib
+    list(APPEND third_party_deps pslib)
+    if(WITH_LIBMCT)
+        include(external/libmct)     # download, build, install libmct
+        list(APPEND third_party_deps libmct)
```
Comment (liupluswei) [inline cmake/third_party.cmake]: seems both "extern_xxx" and "xxx" are used in this file, can you unify the format?

File: cmake/external/pslib_brpc.cmake (hunk 1 above)
Comment (zhwesky2010) [reply cmake/external/pslib_brpc.cmake:1]: I have consult Luotao and copyright is the same with the file when established.

**Review by liupluswei — DISMISSED** [review]
LGTM

**Review by luotao1 — COMMENTED** [review]
100644 → 100755，请不要改文件属性

**Review by luotao1 — APPROVED** [review]
LGTM

**Pull request merged**